README.md
# Linux Terminal

A custom interactive Linux terminal implementation developed in C, utilizing various system calls to provide core terminal functionality.

## 🚀 Features

This terminal supports the following built-in commands:

- **`cd`** - Change directory
- **`mkdir`** - Create directories
- **`rm`** - Remove files and directories
- **`cat`** - Display file contents
- **`history`** - Show command history
- **`echo`** - Display text
- **`exit`** - Exit the terminal
- **`pwd`** - Print working directory

## 🛠️ Technical Implementation

The terminal is built using fundamental Linux system calls:

- **`fork()`** - Create child processes for command execution
- **`getcwd()`** - Get current working directory
- **`wait()`** - Wait for child processes to complete
- **`exec()`** family - Execute commands
- **`chdir()`** - Change directory operations

## 📋 Prerequisites

- Linux operating system
- GCC compiler
- Make utility (optional)

## 🔧 Installation & Usage

### Clone the Repository
```bash
git clone https://github.com/anshmittal0811/Linux-Terminal.git
cd Linux-Terminal
```

### Compile the Program
```bash
gcc -o terminal main.c
```

### Run the Terminal
```bash
./terminal
```

## 💻 Usage Examples

```bash
# Navigate directories
$ pwd
/home/user

$ cd Documents
$ pwd
/home/user/Documents

# Create and manage files/directories
$ mkdir test_folder
$ echo "Hello World" > test.txt
$ cat test.txt
Hello World

# View command history
$ history
1. pwd
2. cd Documents
3. mkdir test_folder
4. echo "Hello World" > test.txt
5. cat test.txt

# Exit the terminal
$ exit
```

## 🏗️ Architecture

The terminal follows a simple architecture:

1. **Command Parser** - Parses user input and identifies commands
2. **Process Management** - Uses `fork()` to create child processes
3. **Built-in Commands** - Implements core commands directly
4. **External Commands** - Executes system programs via `exec()`
5. **History Management** - Maintains command history

## 📝 Supported Command Details

| Command | Description | Example |
|---------|-------------|---------|
| `pwd` | Print current working directory | `pwd` |
| `cd <path>` | Change directory | `cd /home/user` |
| `mkdir <name>` | Create directory | `mkdir newfolder` |
| `rm <file>` | Remove file | `rm file.txt` |
| `cat <file>` | Display file contents | `cat readme.txt` |
| `echo <text>` | Display text | `echo "Hello World"` |
| `history` | Show command history | `history` |
| `exit` | Exit terminal | `exit` |

## 🔍 Implementation Details

### Process Flow
1. Display prompt
2. Read user input
3. Parse command and arguments
4. Check if command is built-in
5. If built-in: execute directly
6. If external: fork() → exec() → wait()
7. Update history
8. Repeat

### Memory Management
- Dynamic memory allocation for command storage
- Proper cleanup of child processes
- History buffer management

## 👨‍💻 Author

**Ansh Mittal** - [anshmittal0811](https://github.com/anshmittal0811)

## 📚 Learning Outcomes

This project demonstrates:
- Understanding of Linux system calls
- Process management in Unix-like systems
- Command-line interface development
- C programming best practices
- Memory management in C

## 🐛 Known Issues

- Limited error handling for edge cases
- Basic command parsing (may not handle complex arguments)
- No support for pipes or redirection

## 🔮 Future Enhancements

- [ ] Add support for command pipes (`|`)
- [ ] Implement input/output redirection (`>`, `<`)
- [ ] Add tab completion
- [ ] Implement job control (background processes)
- [ ] Add more built-in commands (`ls`, `grep`, etc.)
- [ ] Improve error handling and user feedback

---

*This project was created as a learning exercise to understand Linux system programming and terminal implementation.*


history.txt
history 
ls 
pwd 
mkdir at 
clear 
exit 


ls.c
#include<stdio.h>
#include <stdlib.h>
#include<dirent.h>
#include <string.h>
#include <unistd.h>
#define RED "\x1b[31m"
#define YELLOW  "\x1b[33m"
#define RESET   "\x1b[0m"
static int comparator(const void* str1, const void* str2) {
    char *str_1 = *(char**)str1;
    char *str_2 = *(char**)str2;
    char a[1000],b[1000];
    strcpy(a,str_1);
    strcpy(b,str_2);
    for (int i = 0; a[i]!='\0'; i++) {
        if(a[i] >= 'A' && a[i] <= 'Z') {
            a[i] = a[i] + 32;
        }
    }
    for (int i = 0; b[i]!='\0'; i++) {
        if(b[i] >= 'A' && b[i] <= 'Z') {
            b[i] = b[i] + 32;
        }
    }
    if(strcmp(a,b) >= 0)
    {
        return 1;
    }
    else return 0;

}
int main(int argc, char *argv[])
{
    DIR *directory;
    struct dirent *dir;
    char arguments[1000][1000];
    char *arg = strtok(argv[1]," \n");
    int argidx = 0;
    while (arg != NULL)
    {
        strcpy(arguments[argidx++],arg);
        arg = strtok(NULL," \n");
    }
    if(argidx == 0)
    {
        char path_store[1000];
        getcwd(path_store,1000);
        directory = opendir(path_store);
        char *name[10000];
        int index = 0;
        int len = 0;
        if (directory)
        {
            while ((dir = readdir(directory)) != NULL)
            {
                len++;
                name[index++] =  dir->d_name;
            }
            qsort(name, len, sizeof(const char*), comparator);
            for(int i=0; i<len; i++)
            {
                if(*name[i] != 46)
                {
                    printf(YELLOW "%s\n" RESET, name[i]);
                }
            }
            closedir(directory);
        }
    }
    else {
        if (strcmp(arguments[0], "-a") == 0) {
            if (argidx >= 2) {
                for(int i=1; i<argidx; i++) {
                    
                    char path[100];
                    char path_store[1000];
                    char *name[10000];
                    getcwd(path_store, 1000);
                    strcpy(path, path_store);
                    strcat(path, "/");
                    strcat(path, arguments[1]);
                    directory = opendir(path);
                    int index = 0;
                    int len = 0;
                    if (directory) {
                    	 printf("%s:\n",arguments[i]);
                        while ((dir = readdir(directory)) != NULL) {
                            len++;
                            name[index++] = dir->d_name;
                        }
                        qsort(name, len, sizeof(const char *), comparator);
                        for (int j = 0; j < len; j++) {
                            printf(YELLOW "%s\n" RESET, name[j]);
                        }
                        closedir(directory);
                    }
                    else
                    {
                        fprintf(stderr,RED"%s '%s': No such file or directory"RESET,"ls: cannot access",arguments[i]);
                    }
                    printf("\n");
                }
            } else {
                char path_store[1000];
                getcwd(path_store,1000);
                directory = opendir(path_store);
                char *name[10000];
                int index = 0;
                int len = 0;
                if (directory) {
                    while ((dir = readdir(directory)) != NULL) {
                        len++;
                        name[index++] =  dir->d_name;
                    }
                    qsort(name, len, sizeof(const char*), comparator);
                    for(int i=0; i< len; i++)
                    {
                        printf(YELLOW "%s\n" RESET, name[i]);
                    }
                    closedir(directory);
                }
            }
        } else if (strcmp(arguments[0], "-B") == 0) {
            if (argidx >= 2) {
                for(int i=1; i<argidx; i++)
                {
                    char path[100];
                    char path_store[1000];
                    getcwd(path_store,1000);
                    strcpy(path, path_store);
                    strcat(path,"/");
                    strcat(path, arguments[1]);
                    directory = opendir(path);
                    int flag = 0;
                    char *name[10000];
                    int index = 0;
                    int len = 0;
                    if (directory) {
                    	 printf("%s:\n",arguments[i]);
                        while ((dir = readdir(directory)) != NULL) {
                            name[index++] =  dir->d_name;
                            len++;

                        }
                        qsort(name, len, sizeof(const char*), comparator);
                        for(int k=0; k<len; k++)
                        {
                            for (int j = 0; j < strlen(name[k]); j++) {
                                if (*(name[k] + j) == 126) {
                                    flag = 1;
                                    break;
                                }
                            }
                            if (!flag) {
                                printf(YELLOW "%s\n" RESET, name[i]);
                            }
                        }
                        closedir(directory);
                    }
                    else
                    {
                        fprintf(stderr,RED"%s '%s': No such file or directory"RESET,"ls: cannot access",arguments[i]);
                    }
                    printf("\n");
                }
            }
            else {
                char path_store[1000];
                getcwd(path_store,1000);
                directory = opendir(path_store);
                int flag = 0;
                char *name[10000];
                int index = 0;
                int len = 0;
                if (directory) {
                    while ((dir = readdir(directory)) != NULL) {
                        name[index++] =  dir->d_name;
                        len++;
                    }
                    qsort(name, len, sizeof(const char*), comparator);
                    for(int i=0; i<len; i++)
                    {
                        for (int j = 0; j < strlen(name[i]); j++) {
                            if (*(name[i] + j) == 126) {
                                flag = 1;
                                break;
                            }
                        }
                        if (!flag) {
                            printf(YELLOW "%s\n" RESET, name[i]);
                        }
                    }
                    closedir(directory);
                }
            }
        }
        else
        {
            for(int i=0; i<argidx; i++) {
                
                char path[100];
                char path_store[1000];
                getcwd(path_store,1000);
                strcpy(path, path_store);
                strcat(path,"/");
                strcat(path, arguments[i]);
                directory = opendir(path);
                char *name[10000];
                int index = 0;
                int len = 0;
                if (directory) {
                    printf("%s:\n",arguments[i]);
                    while ((dir = readdir(directory)) != NULL) {
                        name[index++] = dir->d_name;
                        len++;
                    }
                    qsort(name, len, sizeof(const char *), comparator);
                    for (int j = 0; j < len; j++) {
                        printf(YELLOW "%s\n" RESET, name[j]);
                    }
                    closedir(directory);
                } else
                {
                    fprintf(stderr,RED"%s '%s': No such file or directory"RESET,"ls: cannot access",arguments[i]);
                }
                printf("\n");
            }
        }
    }
    return(0);
}


rm.c
#include <stdio.h>
#include <string.h>
#define RED "\x1b[31m"
#define RESET "\x1b[0m"
int  main(int argc, char *argv[100000])
{
    char arguments[1000][1000];
    char *arg = strtok(argv[1]," \n");
    int index = 0;
    while (arg != NULL)
    {
        strcpy(arguments[index++],arg);
        arg = strtok(NULL," \n");
    }
    if(index == 0)
    {
        fprintf(stderr,RED"%s\n"RESET,"rm: missing operand");
        return 0;
    }
    if(strcmp(arguments[0],"-v") == 0) {
        if(index == 1)
        {
            fprintf(stderr,RED"%s\n"RESET,"rm: missing operand");
            return 0;
        }
        for (int i = 1; i < index; i++) {
            int error_check = remove(arguments[i]);
            if (error_check != 0) {
                fprintf(stderr,RED"%s"RESET,"rm: cannot remove ");
                fprintf(stderr,RED"'%s' %s\n"RESET, arguments[i], " : No such file or directory");
            }
            else {
                printf("removed '%s'\n", arguments[i]);
            }
        }
    }
    else if(strcmp(arguments[0],"-i") == 0) {
        if(index == 1)
        {
            fprintf(stderr,RED"%s\n"RESET,"rm: missing operand");
            return 0;
        }
        for (int i = 1; i < index; i++) {
            char access;
            printf("rm: remove regular file '%s'? ",arguments[i]);
            scanf("%c",&access);
            if(access == 89 || access == 121) {
                int error_check = remove(arguments[i]);
                if (error_check != 0) {
                    fprintf(stderr,RED"%s"RESET,"rm: cannot remove ");
                	fprintf(stderr,RED"'%s' %s\n"RESET, arguments[i], " : No such file or directory");
                }
            }
        }
    }
    else
    {
        for(int i=0; i<index; i++)
        {
            int error_check = remove(arguments[i]);
            if (error_check != 0) {
                fprintf(stderr,RED"%s"RESET,"rm: cannot remove ");
                fprintf(stderr,RED"'%s' %s\n"RESET, arguments[i], " : No such file or directory");
            }
        }
    }
    return 0;
}


cat.c
#include <stdio.h>
#include <string.h>

int  main(int argc, char *argv[100000])
{
    FILE *fptr;
    char line[100];
    char arguments[1000][1000];
    char *arg = strtok(argv[1]," \n");
    int index = 0;
    while (arg != NULL)
    {
        strcpy(arguments[index++],arg);
        arg = strtok(NULL," \n");
    }
    if(index > 0) {
        if(strcmp(arguments[0],"-n") == 0)
        {
            int counter = 1;
            for (int i = 1; i < index; i++) {
                fptr = fopen(strtok(arguments[i],"\""), "r");
                while (fgets(line, sizeof(line), fptr)) {
                    printf("   %d  ",counter);
                    printf("%s", line);
                    counter++;
                }
            }
        }
        else if(strcmp(arguments[0],"-E") == 0)
        {
            for (int i = 1; i < index; i++) {
                fptr = fopen(strtok(arguments[i],"\""), "r");
                while (fgets(line, sizeof(line), fptr)) {
                    for(int i=0; i<100; i++)
                    {
                        if(line[i] == 0)
                        {
                            line[i-1] = 36;
                            break;
                        }
                    }
                    printf("%s\n", line);
                }
            }
        }
        else
        {
            for (int i = 0; i < index; i++) {
                fptr = fopen(strtok(arguments[i],"\""), "r");
                while (fgets(line, sizeof(line), fptr)) {
                    printf("%s", line);
                }
            }
        }

    }
}


date.c
#include <stdio.h>
#include <string.h>
#include <time.h>
#define RED "\x1b[31m"
#define RESET "\x1b[0m"
int main(int argc, char *argv[100]) {
    time_t rawtime;
    struct tm *info;
    char buffer[100];
    time(&rawtime);
    info = localtime(&rawtime);
    if(argc == 1) {
        strftime(buffer, 100, "%A %d %B %Y %I:%M:%S %p IST", info);
        printf("%s\n", buffer);
    }
    else
    {
        if(strcmp(argv[1],"-I") == 0)
        {
            strftime(buffer, 100, "%Y-%m-%d", info);
            printf("%s\n", buffer);
        }
        else if(strcmp(argv[1],"-R") == 0)
        {
            strftime(buffer, 100, "%a, %d %b %Y %H:%M:%S +0530", info);
            printf("%s\n", buffer);
        }
        else
     	{
     	     fprintf(stderr,RED "%s'%s'\n" RESET,"date: invalid option -- ",argv[1]);
	     return 0;
    	}
    }
    return (0);
}



mkdir.c
#include <stdio.h>
#include <string.h>
#include <sys/stat.h>
#define RED "\x1b[31m"
#define RESET "\x1b[0m"
int  main(int argc, char *argv[100])
{
    char arguments[1000][1000];
    char *arg = strtok(argv[1]," \n");
    int index = 0;
    while (arg != NULL)
    {
        strcpy(arguments[index++],arg);
        arg = strtok(NULL," \n");
    }
    if(index == 0)
    {
        fprintf(stderr,RED"%s\n"RESET,"mkdir: missing operand");
        return 0;
    }
    if(index >= 1) {
        if(strcmp(arguments[0],"-v") == 0) {
        if(index == 1)
	    {
			printf("mkdir: missing operand\n");
			return 0;
	    }
        for(int i=1; i<index; i++) {
            int error_check = mkdir(arguments[i], 0777);
            if (error_check != 0) {
                fprintf(stderr,RED "%s '%s'%s\n"RESET,RED"mkdir: cannot create directory",arguments[i], ": File exists");
            } 
            else {
                printf("mkdir: created directory '%s'\n", arguments[i]);
            }
           }
        }
        else if(strcmp(arguments[0],"--help")) {
            for(int i=0; i<index; i++) {
                int error_check = mkdir(arguments[i], 0777);
                if (error_check != 0) {
                    fprintf(stderr,RED "%s '%s'%s\n"RESET,RED"mkdir: cannot create directory",arguments[i], RED": File exists"RESET);
                }
            }
        }
        else if(strcmp(arguments[0],"--help") == 0)
        {
            printf("Usage: mkdir [OPTION]... DIRECTORY...\n"
                   "Create the DIRECTORY(ies), if they do not already exist.\n"
                   "\n"
                   "Mandatory arguments to long options are mandatory for short options too.\n"
                   "  -m, --mode=MODE   set file mode (as in chmod), not a=rwx - umask\n"
                   "  -p, --parents     no error if existing, make parent directories as needed\n"
                   "  -v, --verbose     print a message for each created directory\n"
                   "  -Z                   set SELinux security context of each created directory\n"
                   "                         to the default type\n"
                   "      --context[=CTX]  like -Z, or if CTX is specified then set the SELinux\n"
                   "                         or SMACK security context to CTX\n"
                   "      --help     display this help and exit\n"
                   "      --version  output version information and exit\n"
                   "\n"
                   "GNU coreutils online help: <https://www.gnu.org/software/coreutils/>\n"
                   "Full documentation at: <https://www.gnu.org/software/coreutils/mkdir>\n"
                   "or available locally via: info '(coreutils) mkdir invocation'\n");
        }
    }
    return 0;
}


shell.c
#include <sys/types.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/wait.h>
#define RED "\x1b[31m"
#define BLUE "\x1b[34m"
#define GREEN "\x1b[32m"
#define CYAN "\x1b[36m"
#define RESET "\x1b[0m"
void history_getter(char *path)
{
    FILE *history;
    char *pathway = (char *)malloc(500*sizeof(char));
    strcpy(pathway,path);
    strcat(pathway,"/history.txt");
    history = fopen(pathway,"r");
    char buffer[1000];
    while(fgets(buffer,1000,history))
    {
        printf("%s",buffer);
    }
    free(pathway);
    fclose(history);
}
void history_setter(char *path, char *string)
{
    FILE *history;
    char *pathway = (char *)malloc(500*sizeof(char));
    strcpy(pathway,path);
    strcat(pathway,"/history.txt");
    history = fopen(pathway,"a+");
    char words[1000];
    strcpy(words,string);
    char *argu = strtok(words," \n");
    while(argu != NULL)
    {
        fprintf(history,"%s ",argu);
        argu = strtok(NULL," \n");
    }
    fprintf(history,"%s", "\n");
    free(pathway);
    fclose(history);
}
void history_clean(char *path)
{
	FILE *history;
    char *pathway = (char *)malloc(500*sizeof(char));
    strcpy(pathway,path);
    strcat(pathway,"/history.txt");
    remove(pathway);
    free(pathway);
}
int main()
{

    char path_store[1000];
    getcwd(path_store,1000);
    printf(BLUE "=====================================================================\n" RESET);
    while (1) {
        printf(CYAN "RUDIMENTARY SHELL ~$ " RESET);
        char str[100];
        fgets(str, 100, stdin);
        history_setter(path_store,str);
        char list[1000][1000];
        char *arguments = strtok(str," \n");
        int index = 0;
        int len = 0;
        while(arguments != NULL)
        {
            len++;
            strcpy(list[index++],arguments);
            arguments = strtok(NULL," \n");
        }
        if (strcmp(list[0], "exit") == 0) {
            break;
        }
        else if (strcmp(list[0], "cd") == 0) {
            if (strcmp(list[1], "..") == 0) {
                chdir("..");
            }
            else
            {
                if(chdir(list[1]) == -1)
                {
                    printf("cd: ");
                    printf("%s",list[1]);
                    printf(": ");
                    printf("No such file or directory\n");
                }
                else
                {
                    chdir(list[1]);
                }
            }
        }
        else if(strcmp(list[0],"history") == 0)
        {
            if(len == 1) {
                history_getter(path_store);
            }
            else if(strcmp(list[1],"-c") == 0)
            {
            	history_clean(path_store);
            }
            else
            {
                printf("history: history [-c] [-d offset] [n] or history -anrw [filename] or history -ps arg [arg...]\n"
                       "    Display or manipulate the history list.\n"
                       "    \n"
                       "    Display the history list with line numbers, prefixing each modified\n"
                       "    entry with a `*'.  An argument of N lists only the last N entries.\n"
                       "    \n"
                       "    Options:\n"
                       "      -c\tclear the history list by deleting all of the entries\n"
                       "      -d offset\tdelete the history entry at position OFFSET. Negative\n"
                       "    \t\toffsets count back from the end of the history list\n"
                       "    \n"
                       "      -a\tappend history lines from this session to the history file\n"
                       "      -n\tread all history lines not already read from the history file\n"
                       "    \t\tand append them to the history list\n"
                       "      -r\tread the history file and append the contents to the history\n"
                       "    \t\tlist\n"
                       "      -w\twrite the current history to the history file\n"
                       "    \n"
                       "      -p\tperform history expansion on each ARG and display the result\n"
                       "    \t\twithout storing it in the history list\n"
                       "      -s\tappend the ARGs to the history list as a single entry\n"
                       "    \n"
                       "    If FILENAME is given, it is used as the history file.  Otherwise,\n"
                       "    if HISTFILE has a value, that is used, else ~/.bash_history.\n"
                       "    \n"
                       "    If the HISTTIMEFORMAT variable is set and not null, its value is used\n"
                       "    as a format string for strftime(3) to print the time stamp associated\n"
                       "    with each displayed history entry.  No time stamps are printed otherwise.\n"
                       "    \n"
                       "    Exit Status:\n"
                       "    Returns success unless an invalid option is given or an error occurs.\n");
            }
        }
        else if(strcmp(list[0],"pwd") == 0)
        {
            char current_store[1000];
            getcwd(current_store,1000);
            printf("%s\n",current_store);
        }
        else if(strcmp(list[0],"echo") == 0)
        {
            if(strcmp(list[1],"-n") == 0)
            {
                int idx = 2;
                while(idx < len)
                {
                    printf("%s ",list[idx]);
                    idx++;
                }
            }
            else if(strcmp(list[1],"-e") == 0)
            {
                int idx = 2;
                while (idx < len)
                {
                    for(int i=0; i<strlen(list[idx]);i++)
                    {
                        if(list[idx][i] - 92 == 0 && list[idx][i+1] - 'n' == 0 && list[2][0] == '"')
                        {
                            printf("\n");
                            i++;
                            continue;
                        }
                        if(list[idx][i] == '"')
                        {}
                        else
                        {
                            if(list[idx][i] - 92 != 0){printf("%c",list[idx][i]);}
                        }
                    }
                    printf("\n");
                    idx++;
                }
            }
            else
            {
                int idx = 1;
                while(idx < len)
                {
                    for(int i=0; i<strlen(list[idx]); i++)
                    {
                        if(list[idx][i] != 34){
                            printf("%c",list[idx][i]);}
                    }
                    idx++;
                }
                printf("\n");
            }
        }
        else if(strcmp(list[0],"date") == 0)
        {
            pid_t pid;
            pid = fork();
            if(pid == 0)
            {
                if(len == 1)
                {
                    strcat(path_store,"/dateo");
                    execl(path_store,"dateo",NULL);
                }
                else if(len == 2)
                {
                    strcat(path_store,"/dateo");
                    execl(path_store,"dateo",list[1],NULL);
                }
            }
            else if(pid < 0)
            {
                fprintf(stderr, RED "Fork failed!\n" RESET);
                exit(-1);
            }
            else
            {
                wait(NULL);
            }
        }
        else if(strcmp(list[0],"rm") == 0)
        {
            pid_t pid;
            pid = fork();
            char file_string[10000];
            strcpy(file_string," ");
            if(pid == 0)
            {
                for(int i=1; i<len; i++)
                {
                    if(i == 1)
                    {
                        strcpy(file_string,list[i]);
                    }
                    else
                    {
                        strcat(file_string,list[i]);
                    }
                    if(i != len-1) {
                        strcat(file_string, " ");
                    }
                    else
                    {
                        strcat(file_string,"\n");
                    }
                }
                printf("%s",file_string);
                strcat(path_store,"/rmo");

                execl(path_store, "rmo", file_string, NULL);
            }
            else if(pid < 0)
            {
                fprintf(stderr, RED "Fork failed!\n" RESET);
                exit(-1);
            }
            else
            {
                wait(NULL);
            }
        }
        else if(strcmp(list[0],"cat") == 0)
        {
            pid_t pid;
            pid = fork();
            char file_string[10000];
            strcpy(file_string," ");
            if(pid == 0)
            {
                for(int i=1; i<len; i++)
                {
                    if(i == 1)
                    {
                        strcpy(file_string,list[i]);
                    }
                    else
                    {
                        strcat(file_string,list[i]);
                    }
                    if(i != len-1) {
                        strcat(file_string, " ");
                    }
                    else
                    {
                        strcat(file_string,"\n");
                    }
                }
                strcat(path_store,"/cato");
                execl(path_store, "cato", file_string, NULL);
            }
            else if(pid < 0)
            {
                fprintf(stderr, RED "Fork failed!\n" RESET);
                exit(-1);
            }
            else
            {
                wait(NULL);
            }
        }
        else if(strcmp(list[0],"ls") == 0)
        {
            pid_t pid;
            pid = fork();

            if(pid == 0)
            {
                char file_string[10000];
                strcpy(file_string," ");
                for(int i=1; i<len; i++)
                {
                    if(i == 1)
                    {
                        strcpy(file_string,list[i]);
                    }
                    else
                    {
                        strcat(file_string,list[i]);
                    }
                    if(i != len-1) {
                        strcat(file_string, " ");
                    }
                    else
                    {
                        strcat(file_string,"\n");
                    }
                }
                strcat(path_store,"/lso");
                execl(path_store, "lso", file_string, NULL);

            }
            else if(pid < 0)
            {
                fprintf(stderr, RED "Fork failed!\n" RESET);
                exit(-1);
            }
            else
            {
                wait(NULL);
            }
        }
        else if(strcmp(list[0],"mkdir") == 0)
        {
            pid_t pid;
            pid = fork();
            char file_string[10000];
            strcpy(file_string," ");
            if(pid == 0)
            {
                for(int i=1; i<len; i++)
                {
                    if(i == 1)
                    {
                        strcpy(file_string,list[i]);
                    }
                    else
                    {
                        strcat(file_string,list[i]);
                    }
                    if(i != len-1) {
                        strcat(file_string, " ");
                    }
                    else
                    {
                        strcat(file_string,"\n");
                    }
                }
                strcat(path_store,"/mkdiro");
                execl(path_store, "mkdiro", file_string, NULL);

            }
            else if(pid < 0)
            {
                fprintf(stderr, RED "Fork failed!\n" RESET);
                exit(-1);
            }
            else
            {
                wait(NULL);
            }
        }
        else
        {
            fprintf(stderr,RED"%s%s\n"RESET,list[0],": command not found");
        }
    }

    printf(BLUE "=====================================================================\n" RESET);
    return 0;
}

